### ISSUE TITLE
revisit DNS Layer and improve

### ISSUE THREAD

**Issue by GlenDC** [plabayo/rama#116]
The DNS Layer is currently more in a PoC state. It kinda works, if you need it, but I think there might be better ways to do some things. In a nutshell this is what the DNS Layer is all about:

- allow layered resolvers (e.g. header overwrite > DNS Map in Mem > System DNS)
- allow DNS Caching of queries
- allow custom resolver to be integrated (e.g. if people want to integrate a resolver provided by stuff like <https://github.com/hickory-dns/hickory-dns>)

Maybe:

- allow DNS Firewall (e.g. some stuff not to resolve), might be not needed given we already have the Hijack middleware which can block on (sub)domains already...

TLDR:

We already have a DNS Layer, but for the v0.3 release might be a nice time to revisit it and see where we might want to improve stuff.

**Comment by GlenDC** [comment]
Going to start the initial work on this already for v0.2 as part of #199.
